File: 2471_minimum_number_of_operations_to_sort_a_binary_tree_by_level.java
public class Solution {
    public int minimumOperations(TreeNode root) {
        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root);
        int result = 0;

        while (!queue.isEmpty()) {
            int size = queue.size();
            List<Integer> level = new ArrayList<>();
            for (int i = 0; i < size; i++) {
                TreeNode node = queue.poll();
                level.add(node.val);
                if (node.left != null) queue.add(node.left);
                if (node.right != null) queue.add(node.right);
            }
            result += countSwaps(level);
        }
        return result;
    }

    private int countSwaps(List<Integer> nums) {
        int n = nums.size(), swaps = 0;
        Integer[] indices = new Integer[n];
        for (int i = 0; i < n; i++) indices[i] = i;

        Arrays.sort(indices, Comparator.comparingInt(nums::get));
        boolean[] visited = new boolean[n];

        for (int i = 0; i < n; i++) {
            if (visited[i] || indices[i] == i) continue;
            int j = i, cycle = 0;
            while (!visited[j]) {
                visited[j] = true;
                j = indices[j];
                cycle++;
            }
            swaps += cycle - 1;
        }
        return swaps;
    }
}

File: 2471_minimum_number_of_operations_to_sort_a_binary_tree_by_level.cpp
class Solution {
public:
    int minimumOperations(TreeNode* root) {
        if (!root) return 0;
        
        queue<TreeNode*> q;
        q.push(root);
        int totalOps = 0;

        while (!q.empty()) {
            int n = q.size();
            vector<int> level;
            
            for (int i = 0; i < n; i++) {
                TreeNode* node = q.front();
                q.pop();
                level.push_back(node->val);
                if (node->left) q.push(node->left);
                if (node->right) q.push(node->right);
            }

            totalOps += minSwapsToSort(level);
        }

        return totalOps;
    }

private:
    int minSwapsToSort(vector<int>& arr) {
        int n = arr.size();
        vector<pair<int, int>> valIdx(n);
        for (int i = 0; i < n; i++) {
            valIdx[i] = {arr[i], i};
        }
        
        sort(valIdx.begin(), valIdx.end());
        vector<bool> visited(n, false);
        int swaps = 0;

        for (int i = 0; i < n; i++) {
            if (visited[i] || valIdx[i].second == i) continue;

            int cycleSize = 0, j = i;
            while (!visited[j]) {
                visited[j] = true;
                j = valIdx[j].second;
                cycleSize++;
            }
            if (cycleSize > 1) swaps += cycleSize - 1;
        }

        return swaps;
    }
};
